speechbrain/lobes/models/beats.py

```diff
@@ -0,0 +1,2024 @@
+
+        # Configuration checks
+        assert not (
+            self.cfg.deep_norm and self.cfg.layer_norm_first
+        ), "Configuration error: 'deep_norm' and 'layer_norm_first' cannot both be True."
+
+        # Initialize encoder and layer normalization
+        self.encoder = TransformerEncoder(self.cfg)
+        self.layer_norm = LayerNorm(self.embed)
+
+        # Define predictor for fine-tuned models
+        if self.cfg.finetuned_model:
+            self.predictor_dropout = nn.Dropout(self.cfg.predictor_dropout)
+            self.predictor = nn.Linear(
+                self.cfg.encoder_embed_dim, self.cfg.predictor_class
+            )
+        else:
+            self.predictor = None
+
+        # Load weights from the checkpoint if available
+        if checkpoint:
+            self.load_state_dict(checkpoint["model"])
+
+        # Set the model to evaluation mode if frozen
+        if self.freeze:
+            self.eval()
+
+    def forward_padding_mask(
+        self, features: torch.Tensor, padding_mask: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Adjusts the padding mask for the given features.
+
+        Arguments
+        ---------
+        features : torch.Tensor
+            Input features after patch embedding.
+        padding_mask : torch.Tensor
+            Original padding mask for input signals.
+
+        Returns
+        -------
+        torch.Tensor
+            Adjusted padding mask.
+        """
+        extra = padding_mask.size(1) % features.size(1)
+        if extra > 0:
+            padding_mask = padding_mask[:, :-extra]
+        padding_mask = padding_mask.view(
+            padding_mask.size(0), features.size(1), -1
+        )
+        return padding_mask.all(-1)
+
+    def preprocess(
+        self,
+        source: torch.Tensor,
+        fbank_mean: float = 15.41663,
+        fbank_std: float = 6.55582,
+    ) -> torch.Tensor:
+        """
+        Preprocesses the input waveform by extracting filter banks and applying normalization.
+
+        Arguments
+        ---------
+        source : torch.Tensor
+            Input waveform signals.
+        fbank_mean : float, optional
+            Mean value for filter bank normalization (default: 15.41663).
+        fbank_std : float, optional
+            Standard deviation for filter bank normalization (default: 6.55582).
+
+        Returns
+        -------
+        torch.Tensor
+            Normalized filter banks.
+        """
+        fbanks = []
+        for waveform in source:
+            waveform = waveform.unsqueeze(0) * 2**15
+            fbank = ta_kaldi.fbank(
+                waveform,
+                num_mel_bins=128,
+                sample_frequency=16000,
+                frame_length=25,
+                frame_shift=10,
+            )
+            fbanks.append(fbank)
+        fbank = torch.stack(fbanks, dim=0)
+        return (fbank - fbank_mean) / (2 * fbank_std)
+
+    def extract_features(
```

Comment:
It would be good to also have a `forward` function that wraps `extract_features` as done in `huggingface_transformers/wav2vec2.py`